relo_core: adjust targ_spec member walk

rafaeldtinoco opened the pull request:
- fix typedef relocation member btf_type's
- fix null dereference when relo->kind = BPF_TYPE_ID_LOCAL
- add btf_reloc_dump_spec to facilitate development

rafaeldtinoco committed:
relo_core: adjust targ_spec member walk

Closes: #4

- fix typedef relocation member btf_type's
- fix null dereference when relo->kind = BPF_TYPE_ID_LOCAL
- add btf_reloc_dump_spec to facilitate development

rafaeldtinoco committed:
relo_core: split different types of relocations

rafaeldtinoco committed:
relo_core: better organize btf_reloc_add_type

- easy spot if missing (to be supported) complex relocation types

rafaeldtinoco committed:
relo_core: adjust typedefs resolution

Fixes: #12

rafaeldtinoco commented:
These patches are on top of:

```
commit 9e61ce8
Merge: be7c4a6 c2c3cdb
Author: Mauricio Vásquez <mauricio@kinvolk.io>
Date:   Tue Oct 5 17:33:25 2021

    Merge pull request #3 from kinvolk/mauricio/make-struct-info-private

    Make struct reloc_info private
```

I confess I had problems with the linked list replacement (stack smashes and seg faults). I did not look the hashmap functions to debug, so far. Might do a rebase and check again next week, but would appreciate feedback.

Thanks

mauriciovasquezbernal committed:
change api to use new btf_reloc_get_type()

mauriciovasquezbernal committed:
remove uneeded resolve typedefs logic

mauriciovasquezbernal commented:
> I confess I had problems with the linked list replacement (stack smashes and seg faults). I did not look the hashmap functions to debug, so far. Might do a rebase and check again next week, but would appreciate feedback.
 
I rebased it for you https://github.com/kinvolk/libbpf/tree/mauricio/rafael-wip-rebased-to-hashmap. I tried it and it seems to work fine (trying with the portablebpf program). 

I'll review this PR.

mauriciovasquezbernal commented on src/relo_core.c:
```diff
@@ -1472,39 +1485,116 @@ struct btf *bpf_reloc_info_get_btf(struct btf_reloc_info *info) {
-			// set the current type to array type and continue processing
-			reloc_type = btf_reloc_get_type(info, array_type);
 
+		if (!targ_spec->spec[i].name)
 			continue;
```
What's the reason to skip in this case? If the name is `NULL` it means it refers to an array access, I think if you skip it when some types will be missing from the generated BTF.

mauriciovasquezbernal replied:
btw, I have to remove it on my rebase to make it work, otherwise I get a segfault.

rafaeldtinoco replied:
Yep. This was a leftover from attempts before btf_is_mod() and btf_is_typedef() were checked. Will fix it.

rafaeldtinoco replied:
@mauriciovasquezbernal BTW, this change made the missing types I needed in order to run tracee. I was able to run tracee for the first time now =)... Generated BTF file is 5k and all relocations were satisfied for tracee:

```
$ sudo TRACEE_BTF_FILE=/home/rafaeldtinoco/work/sources/ebpf/btfgen/generated/prefix-5.4.0-84-generic.btf ./dist/tracee-ebpf --debug --trace event=execve --trace event=execveat
OSInfo: VERSION: "18.04.6 LTS (Bionic Beaver)"
OSInfo: ID: ubuntu
OSInfo: ID_LIKE: debian
OSInfo: PRETTY_NAME: "Ubuntu 18.04.6 LTS"
OSInfo: VERSION_ID: "18.04"
OSInfo: VERSION_CODENAME: bionic
OSInfo: KERNEL_RELEASE: 5.4.0-84-generic
BTF: bpfenv = false, btfenv = true, vmlinux = false
BPF: using embedded BPF object
BTF: using BTF file from environment: /home/rafaeldtinoco/work/sources/ebpf/btfgen/generated/prefix-5.4.0-84-generic.btf
unpacked CO:RE bpf object file into memory
TIME             UID    COMM             PID     TID     RET              EVENT                ARGS
04:53:21:212781  1000   bash             13207   13207   0                execve               pathname: /bin/ls, argv: [ls --color=auto]
04:53:21:234339  1000   bash             13212   13212   0                execve               pathname: /usr/bin/git, argv: [git branch]
04:53:21:234981  1000   bash             13213   13213   0                execve               pathname: /bin/sed, argv: [sed -e /^[^*]/d -e s/* \(.*\)/\1/]

End of events stream
Stats: {EventCount:3 ErrorCount:0 LostEvCount:0 LostWrCount:0 LostNtCount:0}
```

wanted to share the good news.

mauriciovasquezbernal commented on src/relo_core.c:
```diff
@@ -1472,39 +1485,116 @@ struct btf *bpf_reloc_info_get_btf(struct btf_reloc_info *info) {
+		if (!targ_spec->spec[i].name)
 			continue;
+		else {
+			while (btf_is_mod(btf_type) || btf_is_typedef(btf_type)) {
+				btf_type = (struct btf_type*) btf__type_by_id(btf, btf_type->type);
+				reloc_type = btf_reloc_get_type(info, btf_type);
```
What about doing it on `case BTF_KIND_TYPEDEF`?
Do you know/think there is any case where `btf_is_mod()` would be true?

rafaeldtinoco replied:
A volatile inside a struct:

```
[80] STRUCT 'sock_common' size=187 vlen=6
        'skc_family' type_id=81 bits_offset=128
        '(anon)' type_id=83 bits_offset=0
        '(anon)' type_id=86 bits_offset=96
        'skc_state' type_id=98 bits_offset=144
        'skc_v6_rcv_saddr' type_id=100 bits_offset=576
        'skc_v6_daddr' type_id=100 bits_offset=448
```

type 98 and 118: 

```
[98] VOLATILE '(anon)' type_id=118
[118] INT 'unsigned char' size=1 bits_offset=0 nr_bits=8 encoding=(none)
```

It gets added by its own relocation speculation (btf_reloc_add_type) BUT it needs to be skipped when the relocation that contains it as a member is reached.

Unless I did not understand this correctly =o)

rafaeldtinoco replied:
Its easier to do it before because we need to skip typedef of typedefs as well :\. This loop is actually taken from the skip_mod_and_typedef (or something like it). There is a function with that approach BUT we needed the reloc to be redefined each resolution.. so I wrote it again instead of using it.

This while the while loop solves all the mods/typedefs before moving on into the add type logic.

mauriciovasquezbernal replied:
> A volatile inside a struct:

You're right. 

> It gets added by its own relocation speculation (btf_reloc_add_type)

I think it's actually missing -> https://github.com/kinvolk/libbpf/pull/11#discussion_r726567954. 

> skipped when the relocation that contains it as a member is reached.

yeah, I totally agree. 

I think the logic we should follow is: 
- in `btf_reloc_add_type` we add the type and all the types it depends on (we resolve all typedefs, "mods", array type and array index type, etc). 
- In `btf_reloc_info_gen_field` we skip the types until we get the type we need (struct, union, array, ...).

mauriciovasquezbernal commented on src/relo_core.c:
```diff
@@ -1472,39 +1485,116 @@ struct btf *bpf_reloc_info_get_btf(struct btf_reloc_info *info) {
+			this = (struct btf_type *)btf__type_by_id(btf, a->type);
+			reloc_type = btf_reloc_get_type(info, this);
+			continue;
+		//case BTF_KIND_TYPEDEF:
+		//	reloc_type = btf_reloc_get_type(info, btf_type);
+		//	btf_type = (struct btf_type*) btf__type_by_id(btf, btf_type->type);
+		//	continue;
```
To make it work you need to:

1. `break` instead of `continue`: After solving the typedef we still want the logic below to be executed to add the members for this type. 
2. `reloc_type = btf_reloc_get_type(info, btf_type);` ->  `reloc_type = btf_reloc_get_type(info, btf_type->type);` We want `reloc_type` to be the anon struct and not the typedef. (too add the members to the anon struct below). 


```suggestion
		//case BTF_KIND_TYPEDEF:
		//	reloc_type = btf_reloc_get_type(info, btf_type->type);
		//	btf_type = (struct btf_type*) btf__type_by_id(btf, btf_type->type);
		//	break;
```

rafaeldtinoco replied:
Check my previous answer, I think it covers this case. It wont work here because we need the full typedef/mods resolution, like the **skip_mods_or_typedefs** function (but considering reloc re-definition for our logic).

mauriciovasquezbernal replied:
yeah, you're right. It's better to skip them at the beginning.

mauriciovasquezbernal commented on src/relo_core.c:
```diff
@@ -1472,39 +1485,116 @@ struct btf *bpf_reloc_info_get_btf(struct btf_reloc_info *info) {
+		switch (btf_kind(btf_type)) {
+		case BTF_KIND_STRUCT:
+		case BTF_KIND_UNION:
+			break;
```
What about putting all the struct / union specific code that we have at the end in this case stament?

rafaeldtinoco replied:
Hum.. yep, let me play with it... lets see..

mauriciovasquezbernal commented on src/relo_core.c:
```diff
@@ -1732,7 +1847,7 @@ int bpf_core_apply_relo_insn(const char *prog_name, struct bpf_insn *insn,
 
 patch_insn:
-	if (reloc_info) {
+	if (reloc_info && targ_res.targ_spec) {
```
Did you hit any case where `targ_res.targ_spec` was null? I'd like to know when this happens.

rafaeldtinoco replied:
I did. For local relocations. Check in the same function, in the beginning. There is a jump to patch_insn.

mauriciovasquezbernal reviewed (commented):
I think these changes are very good. I like the reorganization of the code to be able to handle other cases. 

Some comments on the code.

mauriciovasquezbernal commented on src/relo_core.c:
```diff
@@ -1234,55 +1234,68 @@ static void btf_reloc_append_type(struct btf_reloc_info *info, struct btf_reloc_
+	case BTF_KIND_INT:
+	case BTF_KIND_PTR:
+	case BTF_KIND_STRUCT:
+	case BTF_KIND_UNION:
+	case BTF_KIND_VOLATILE:
+		break;
+	case BTF_KIND_TYPEDEF:
```
```suggestion
	case BTF_KIND_UNION:
		break;
	case BTF_KIND_VOLATILE:
	case BTF_KIND_TYPEDEF:
```

Otherwise the volatile's type won't be added on the generated BTF.

fntlnz commented on src/relo_core.c:
```diff
@@ -1732,7 +1847,7 @@ int bpf_core_apply_relo_insn(const char *prog_name, struct bpf_insn *insn,
 
 patch_insn:
-	if (reloc_info) {
+	if (reloc_info && targ_res.targ_spec) {
```
```suggestion
	if (reloc_info && targ_res.validate && targ_res.targ_spec) {
```

since targ_spec is a pointer the compiler has no guarantees it will be able to find targ_res initialized after a goto because targ_res itself is a raw value.

There are two ways of solving it
-  checking a scalar value that has a default value (like `validate` in this case)
- make targ_res a pointer

I sent a suggestion here with the first solution

mauriciovasquezbernal replied:
What about doing `targ_res.targ_spec = NULL;` just before `goto patch_insn;`?

rafaeldtinoco replied:
Yep, I thought the same (on the = NULL) when I first read @fntlnz's comment. I think its cleaner.

fntlnz commented on src/relo_core.c:
```diff
@@ -1472,39 +1485,116 @@ struct btf *bpf_reloc_info_get_btf(struct btf_reloc_info *info) {
+	t = btf__type_by_id(spec->btf, type_id);
+	s = btf__name_by_offset(spec->btf, t->name_off);
+
+	printf("RELOCATION: [%u] %s %s", type_id, btf_kind_str(t), str_is_empty(s) ? "<anon>" : s);
```
This is very useful to have printed out

fntlnz commented on src/relo_core.c:
```diff
@@ -1234,55 +1234,68 @@ static void btf_reloc_append_type(struct btf_reloc_info *info, struct btf_reloc_
+	unsigned int id = reloc_type->type->type;
 
+	switch (btf_kind(reloc_type->type)) {
+	case BTF_KIND_INT:
```
```suggestion
	case BTF_KIND_INT:
	case BTF_KIND_FLOAT:
```

fntlnz reviewed (commented):
code LGTM, a couple of comments

Good job @rafaeldtinoco

rafaeldtinoco committed:
relo_core: allow unamed member types to be added as reloc types

rafaeldtinoco committed:
relo_core: volatile and typedef are handled similarly when being added

rafaeldtinoco committed:
relo_core: make sure compiler tracks targ_spec as NULL correctly

rafaeldtinoco committed:
relo_core: reorganize btf_reloc_info_gen_field switch block

rafaeldtinoco commented:
Okay @mauriciovasquezbernal. This is version on top of linked list and it is working very well with tracee with these changes. 1/2 of them you already rebased for the hashlist, I'll continue that work now and propose it here. Just wanted a fully working branch for my tracee's work.

mauriciovasquezbernal committed:
Fix some issues

- don't crash if a structure/union doesn't have any member
- fail if btf_reloc_add_type fails
- check if type for typdef existing before adding a new one
- fail it adding a member to a struct/union fails

mauriciovasquezbernal committed:
Fix btf_reloc_id_get()

We're using a hashmap to store the ids, in order to avoid storing a
pointer to an integer we just store the integer directly into the hasmap
value by using uint_as_hash_key().

The hashmap saves pointers, then the variable we're going to store
the value in must have the right size. This commit uses uintptr_t
to avoid a stack corruption.

mauriciovasquezbernal committed:
Print warning if btf_reloc_id_get fails

TODO: this function should return a boolean instead

rafaeldtinoco committed:
relo_core: treat CONST and PTR like other modifier types

mauriciovasquezbernal committed:
Don't add btf_void type

btf_void is a special type (void) with id 0. It's implicit in the BTF
so don't add it.

rafaeldtinoco committed:
relo_core: BTF_KIND_VOID newid is also 0

- comment third pass as it was breaking final BTF

mauriciovasquezbernal commented:
There are still so many things to improve but I'm merging it to have a common base to work on.
